### PR TITLE
New-DbaDbUser: Refactor and added contained user creation

### LINKED ISSUE #8815: New-DbaDbUser - Contained user
### Summarize Functionality

The cmdlet New-DbaDbUser supports creating contained users, but only Windows users - like this:
`New-DbaDbUser -SqlInstance "SQLServer" -Database "SQLAutomationStore" -Username "domain\samaccountname"`

But we are missing a way to add a SQL user, with a normal username & password - the workaround for me, have been this:
```
Invoke-DbaQuery -SqlInstance "SQLServer" -Database "SQLAutomationStore" -Query "
CREATE USER [$($LoginName)] WITH PASSWORD='$($LoginPassword)'
GO
ALTER ROLE [$($LoginPermission)] ADD MEMBER [$($LoginName)]
GO
"
```

### Is there a command that is similiar or close to what you are looking for?

Yes

### Technical Details

_No response_

### PR DESCRIPTION
**Pull request by sqlarticles**
## Type of Change
 - [ ] Bug fix (non-breaking change, fixes # )
 - [x] New feature (non-breaking change, adds functionality, fixes #8815
 - [ ] Breaking change (affects multiple commands or functionality, fixes # )
 - [x] Ran manual Pester test and has passed (`.\tests\manual.pester.ps1`)
 - [ ] Adding code coverage to existing functionality
 - [x] Pester test is included
 - [ ] If new file reference added for test, has is been added to github.com/dataplat/appveyor-lab ?
 - [ ] Unit test is included
 - [ ] Documentation
 - [ ] Build system

### Purpose
- Refactored the code a bit to use `dbatools` public functions.
- Added parameter sets based on the type of user
- Added functionality to create contained users
- Added functionality to return an warning when the given database doesn't exists
- Removed deprecated `SqlLogin` user type
- Updated integration test

### Approach
Included a new parameter (Password) to the function and also used parameter sets. Depends on which parameter the user choose then action will be taken. In this case if they use `Password` parameter then it will create a SQL user with password in the database. SQL User with password is supported only in contained database and from SQL 2012 onwards. This check is not done , if the database is not enabled partial containment then at the time of execution a error will be thrown.

### Commands to test
`New-DbaDbUser -SqlInstance sqlserver2016 -Database DB1 -Username user1 -Password (ConvertTo-SecureString -String "DBATools" -AsPlainText)`

### CODE REVIEW

**Comment by sqlarticles** [comment]
There are few other tests failing. Do I need to rectify those?

File: public/New-DbaDbUser.ps1 [hunk 1]
```diff
@@ -28,9 +28,12 @@ function New-DbaDbUser {
     .PARAMETER Login
         When specified, the user will be associated to this SQL login and have the same name as the Login.
 
-    .PARAMETER Username
+    .PARAMETER UserName
```
Comment (wsmelton) [inline public/New-DbaDbUser.ps1:31]: Please revert, this is improper spelling/casing of the word.

File: public/New-DbaDbUser.ps1 [hunk 2]
```diff
@@ -88,136 +91,127 @@ function New-DbaDbUser {
...
+        [switch]$IncludeSystem = $False,
+        [Parameter(ParameterSetName = "NoLogin", Mandatory = $False)]
+        [Parameter(ParameterSetName = "Login", Mandatory = $True)]
         [string]$Login,
-        [string]$Username = $Login,
+        [Parameter(ParameterSetName = "Login", Mandatory = $False)]
+        [Parameter(ParameterSetName = "NoLogin", Mandatory = $True)]
+        [Parameter(ParameterSetName = "ContainedSQLUser", Mandatory = $True)]
+        [Parameter(ParameterSetName = "ContainedAADUser", Mandatory = $True)]
+        [string]$UserName,
```
Comment (wsmelton) [inline public/New-DbaDbUser.ps1:114]: Why remove the default value? This is a big breaking change that should not be required.

File: public/New-DbaDbUser.ps1 [hunk 3]
```diff
@@ -88,136 +91,127 @@ function New-DbaDbUser {
...
-            } else {
-                Stop-Function -Message "User $($User.Name) already exists in the database $($User.Parent.Name) on $instance and -Force was not specified, skipping." -Target $User -Continue
-            }
+        # When user is created from login and no user name is provided then login name will be used as the user name
+        if ($Login -and -not($UserName)) {
+            $UserName = $Login
+        }
+
+        #Set appropriate user type
+        #Removed SQLLogin  user type. This is deprecated and the alternate is to map to SqlUser. Reference https://learn.microsoft.com/en-us/dotnet/api/microsoft.sqlserver.management.smo.usertype?view=sql-smo-160
```
Comment (wsmelton) [inline public/New-DbaDbUser.ps1:137]: Your link does not show that is deprecated. This will still work on lower versions of SQL Server?

File: public/New-DbaDbUser.ps1 [hunk 4]
```diff
@@ -88,136 +91,127 @@ function New-DbaDbUser {
...
-                $databases = $databases | Where-Object Name -In $Database
-            } else {
-                if (-not $IncludeSystem) {
-                    $databases = $databases | Where-Object IsSystemObject -ne $true
-                }
-            }
-            if ($ExcludeDatabase) {
-                $databases = $databases | Where-Object Name -NotIn $ExcludeDatabase
-            }
+            $databases = Get-DbaDatabase @getDbParam
```
Comment (wsmelton) [inline public/New-DbaDbUser.ps1:166]: We tend to not use this in our commands internally because of performance impact. I would prefer this to be reverted to the original pattern as that is what is followed in most all of our commands.

**Comment by wsmelton** [comment]
The test failures are related to our tests use New-DbaDbUser so yes those need to be addressed.

**Comment by potatoqualitee** [comment]
hey cool! thank you for the PR. I plan to return to dbatools this weekend and will take a closer look as well.

File: public/New-DbaDbUser.ps1 (hunk 1 above)
Comment (sqlarticles) [reply public/New-DbaDbUser.ps1:31]: Thanks for spotting this, It's slipped through

Comment (sqlarticles) [reply public/New-DbaDbUser.ps1:31]: It was mentioned in the fields section.  https://learn.microsoft.com/en-us/dotnet/api/microsoft.sqlserver.management.smo.usertype?view=sql-smo-160#fields

`SqlLogin	0	Specifies a SQLLogin user. This is deprecated, use SQLUser instead.`

I might be wrong. I thought the smo library is part of `dbatools` or `dbatools.library` module as a dependency which should be the recent version. If this is the case then I will prefer to keep the changes. However, if the module uses smo library installed locally then it makes sense to revert it back. I was not clear, your inputs will be helpful.

File: public/New-DbaDbUser.ps1 (hunk 3 above)
Comment (sqlarticles) [reply public/New-DbaDbUser.ps1:136]: It is mentioned in the fields section.  https://learn.microsoft.com/en-us/dotnet/api/microsoft.sqlserver.management.smo.usertype?view=sql-smo-160#fields

`SqlLogin	0	Specifies a SQLLogin user. This is deprecated, use SQLUser instead.`

I might be wrong. I thought the smo library is part of `dbatools` or `dbatools.library` module as a dependency which should be the recent version. If this is the case then I will prefer to keep the changes. However, if the module uses smo library installed locally then it makes sense to revert it back. I was not clear, your inputs will be helpful.

File: public/New-DbaDbUser.ps1 (hunk 1 above)
Comment (sqlarticles) [reply public/New-DbaDbUser.ps1:31]: Thanks for spotting, it slipped through.

File: public/New-DbaDbUser.ps1 (hunk 4 above)
Comment (sqlarticles) [reply public/New-DbaDbUser.ps1:165]: Sorry I didnt get it. You meant the hash table as parameters or using the `Get-DbaDatabase` function?

**Comment by sqlarticles** [comment]
> The test failures are related to our tests use `New-DbaDbUser` so those need to be addressed.

Fixed the failing tests and all the tests are  successful now.

**Comment by sqlarticles** [comment]
> Please use the template provided for the pull request when it deals with modifying code. It helps in review of the PR.

Thanks for this, updated the description using the template. My first PR with dbatools so please bear with me till I get used to it.

File: public/New-DbaDbUser.ps1 (hunk 4 above)
Comment (wsmelton) [reply public/New-DbaDbUser.ps1:165]: My request is reverting the chnage back to the original pattern. `Get-DbaDatabase` incurs performance on many systems and use of that in any command can cause degradation.

File: public/New-DbaDbUser.ps1 [hunk 5]
```diff
@@ -88,136 +91,132 @@ function New-DbaDbUser {
...
 
+    .EXAMPLE
+        PS C:\> New-DbaDbUser -SqlInstance sqlserver2014 -Database DB1 -Username user1 -Password (ConvertTo-SecureString -String "DBATools" -AsPlainText)
+
+        Creates a new cointained sql user named user1 in the database given database with the password specified.
     #>
     [CmdletBinding(SupportsShouldProcess, ConfirmImpact = "Medium")]
     param(
-        [parameter(Mandatory, Position = 1)]
+        [parameter(Mandatory = $True, Position = 1)]
```
Comment (wsmelton) [inline public/New-DbaDbUser.ps1:101]: Please revert this change, PowerShell does not require to explicitly set the true variable.
```suggestion
        [parameter(Mandatory, Position = 1)]
```

File: public/New-DbaDbUser.ps1 [hunk 6]
```diff
@@ -88,136 +91,132 @@ function New-DbaDbUser {
...
         [DbaInstanceParameter[]]$SqlInstance,
         [PSCredential]$SqlCredential,
         [string[]]$Database,
         [string[]]$ExcludeDatabase,
-        [switch]$IncludeSystem,
+        [switch]$IncludeSystem = $False,
+        [Parameter(ParameterSetName = "Login", Mandatory = $True)]
```
Comment (wsmelton) [inline public/New-DbaDbUser.ps1:107]:
```suggestion
        [Parameter(Mandatory, ParameterSetName = "Login")]
```

File: public/New-DbaDbUser.ps1 [hunk 7]
```diff
@@ -88,136 +91,132 @@ function New-DbaDbUser {
...
         [string]$Login,
-        [string]$Username = $Login,
+        [Parameter(ParameterSetName = "Login", Mandatory = $False)]
+        [Parameter(ParameterSetName = "NoLogin", Mandatory = $True)]
```
Comment (wsmelton) [inline public/New-DbaDbUser.ps1:110]:
```suggestion
        [Parameter(Mandatory, ParameterSetName = "NoLogin")]
```

File: public/New-DbaDbUser.ps1 [hunk 8]
```diff
@@ -88,136 +91,132 @@ function New-DbaDbUser {
...
         [switch]$ExternalProvider,
         [switch]$Force,
         [switch]$EnableException
     )
 
     begin {
+        $connParam = @{ }
+        if ($SqlCredential) { $connParam.SqlCredential = $SqlCredential }
+
+        # is this required?
```
Comment (wsmelton) [inline public/New-DbaDbUser.ps1:127]: Please remove comment, it is required.

**Comment by potatoqualitee** [comment]
Thank you for the proposed changes, @wsmelton  🙇🏼  I have merged them.

**Comment by potatoqualitee** [comment]
sweet, the merges ultimately resulted in a pass.

File: public/New-DbaDbUser.ps1 (hunk 4 above)
Comment (sqlarticles) [reply public/New-DbaDbUser.ps1:164]: Gotcha. Before making this change I raised in the Slack channel and I was told there isn't any standard and suggested to use functions if there is one already exists. That's when I refactored to use the functions instead.

I'm happy to revert the changes to use SMO connection instead of the function. Since this PR is merged, do you guys want me to revert it and raise a new PR?

Also, this is my understanding. These are operational functions and not a real time application so it should be fine even if it add few ms to the execution. I know this module is used in wide scope so I dont want to mess up with the expectation. Like I said before, let me know if I need to revert it and create a new PR?

File: public/New-DbaDbUser.ps1 (hunk 7 above)
Comment (sqlarticles) [reply public/New-DbaDbUser.ps1]: I prefer to keep it declared so that it is easy for any newbie like me to understand it clearly

File: public/New-DbaDbUser.ps1 (hunk 8 above)
Comment (sqlarticles) [reply public/New-DbaDbUser.ps1]: shame! I kept that comment. Apologies.
